mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java

```diff
@@ -0,0 +1,322 @@
+               .orElseGet(() -> boundStatement.setToNull(TEXTUAL_LINE_COUNT));
+    }
+
+    private UDTValue toUDT(MessageAttachment messageAttachment) {
+        return typesProvider.getDefinedUserType(ATTACHMENTS)
+            .newValue()
+            .setString(Attachments.ID, messageAttachment.getAttachmentId().getId())
+            .setString(Attachments.NAME, messageAttachment.getName().orNull())
+            .setString(Attachments.CID, messageAttachment.getCid().transform(Cid::getValue).orNull())
+            .setBool(Attachments.IS_INLINE, messageAttachment.isInline());
+    }
+
+    private ByteBuffer toByteBuffer(InputStream stream) throws IOException {
+        return ByteBuffer.wrap(ByteStreams.toByteArray(stream));
+    }
+    
+    public CompletableFuture<Stream<Pair<MailboxMessage, Stream<MessageAttachmentById>>>> retrieveMessages(List<ComposedMessageIdWithMetaData> messageIds, FetchType fetchType, Optional<Integer> limit) {
+        return retrieveRows(messageIds, fetchType, limit)
+                .thenApply(resultSet -> CassandraUtils.convertToStream(resultSet)
+                        .map(row -> message(row, messageIds, fetchType)));
+    }
+
+    private CompletableFuture<ResultSet> retrieveRows(List<ComposedMessageIdWithMetaData> messageIds, FetchType fetchType, Optional<Integer> limit) {
+        return cassandraAsyncExecutor.execute(
+                buildSelectQueryWithLimit(
+                        buildQuery(messageIds, fetchType), 
+                        limit)
+                );
+    }
+    
+    private Where buildQuery(List<ComposedMessageIdWithMetaData> messageIds, FetchType fetchType) {
+        return select(retrieveFields(fetchType))
+                .from(TABLE_NAME)
+                .where(in(MESSAGE_ID, messageIds.stream()
+                        .map(ComposedMessageIdWithMetaData::getComposedMessageId)
+                        .map(ComposedMessageId::getMessageId)
+                        .map(messageId -> (CassandraMessageId) messageId)
+                        .map(CassandraMessageId::get)
+                        .collect(Collectors.toList())));
+    }
+
+    private Pair<MailboxMessage, Stream<MessageAttachmentById>> message(Row row, List<ComposedMessageIdWithMetaData> messageIds, FetchType fetchType) {
+        try {
+            ComposedMessageIdWithMetaData messageIdWithMetaData = retrieveComposedMessageId(messageIdFactory.of(row.getUUID(MESSAGE_ID)), messageIds);
+            ComposedMessageId messageId = messageIdWithMetaData.getComposedMessageId();
+
+            SimpleMailboxMessage message =
+                    new SimpleMailboxMessage(
+                            messageId.getMessageId(),
+                            row.getDate(INTERNAL_DATE),
+                            row.getLong(FULL_CONTENT_OCTETS),
+                            row.getInt(BODY_START_OCTET),
+                            buildContent(row, fetchType),
+                            messageIdWithMetaData.getFlags(),
```

Comment:
it's not ok, why do you need data from the outside ? It would mean you don't return the right type ?